vm/src/objbool.rs

```diff
@@ -1,10 +1,41 @@
-use super::pyobject::{PyObjectKind, PyObjectRef};
+use super::objtype;
+use super::pyobject::{
+    AttributeProtocol, PyContext, PyFuncArgs, PyObjectKind, PyObjectRef, PyResult,
+};
+use super::vm::VirtualMachine;
 
-pub fn boolval(o: PyObjectRef) -> bool {
-    let obj = o.borrow();
-    match obj.kind {
+pub fn boolval(vm: &mut VirtualMachine, obj: PyObjectRef) -> Result<bool, PyObjectRef> {
+    let result = match obj.borrow().kind {
         PyObjectKind::Boolean { value } => value,
         PyObjectKind::Integer { value } => value != 0,
-        ref kind => unimplemented!("converting to boolean unsupported for: {:?}", kind),
+        PyObjectKind::Float { value } => value != 0.0,
+        PyObjectKind::List { ref elements } => !elements.is_empty(),
```

Comment:
Apparently not via ```.__bool__()```

```
$ python3 -c '[].__bool__()'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
AttributeError: 'list' object has no attribute '__bool__'
```

Still not sure where this should live, and this seemed simplest. A trait on PyObject or PyObjectRef is how we do strings and attribute access. But they're still grouped together, so I'm not sure there's an advantage.

If PyObjectKind was a trait rather than enum we could put it on the individual types, but that's a bigger change.

Reply:
Oddly int and float do define ```__bool__```. str, dict, tuple and type don't. Even object doesn't.

Reply:
Would it make sense to give each object a `__bool__`, `__int__`, `__float__`, `__complex__` type?